pkg/reconciler/ingress/status.go

```diff
@@ -0,0 +1,301 @@
+			if probeHost != "" {
+				return false, fmt.Errorf("only one probe host can be defined in a VirtualService, found at least 2: %s, %s", probeHost, host)
+			}
+			probeHost = host
+		}
+	}
+	if probeHost == "" {
+		m.logger.Errorf("the provided VirtualService doesn't contain a probe host. Hosts: %v", vs.Spec.Hosts)
+		return false, errors.New("only a VirtualService with a probe host can be probed")
+	}
+
+	ready, ok := func() (bool, bool) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		if state, ok := m.probingStates[key]; ok {
+			if state.probeHost == probeHost {
+				state.lastAccessed = time.Now()
+				return atomic.LoadInt32(&state.pendingCount) == 0, true
+			}
+
+			// Cancel the polling for the outdated version
+			state.cancel()
+			delete(m.probingStates, key)
+		}
+		return false, false
+	}()
+	if ok {
+		return ready, nil
+	}
+
+	podIPs, err := m.listVirtualServicePodIPs(vs)
+	if err != nil {
+		return false, fmt.Errorf("failed to list the IP addresses of the Pods impacted by the VirtualService: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	state := &probingState{
+		virtualService: vs,
+		probeHost:      probeHost,
+		pendingCount:   int32(len(podIPs)),
+		lastAccessed:   time.Now(),
+		context:        ctx,
+		cancel:         cancel,
+	}
+
+	func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		m.probingStates[key] = state
+	}()
+	for _, podIP := range podIPs {
+		workItem := &workItem{
+			probingState: state,
+			podIP:        podIP,
+		}
+		m.workQueue.AddRateLimited(workItem)
+	}
+
+	return len(podIPs) == 0, nil
+}
+
+// Start starts the StatusManager background operations
+func (m *StatusProber) Start(done <-chan struct{}) {
+	// Start the worker goroutines
+	for i := 0; i < m.probeConcurrency; i++ {
+		go func() {
+			for m.processWorkItem() {
+			}
+		}()
+	}
+
+	// Cleanup the states periodically
+	go func() {
+		wait.Until(m.expireOldStates, m.cleanupPeriod, done)
+	}()
+
+	// Stop processing the queue when cancelled
+	go func() {
+		<-done
+		m.workQueue.ShutDown()
+	}()
+}
+
+// expireOldStates removes the states that haven't been accessed in a while.
+func (m *StatusProber) expireOldStates() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for key, state := range m.probingStates {
+		if time.Since(state.lastAccessed) > m.stateExpiration {
+			state.cancel()
+			delete(m.probingStates, key)
+		}
+	}
+}
+
+// processWorkItem processes a single work item from workQueue.
+// It returns false when there is no more items to process, true otherwise.
+func (m *StatusProber) processWorkItem() bool {
+	obj, shutdown := m.workQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	defer m.workQueue.Done(obj)
+
+	// Crash if the item is not of the expected type
+	item, ok := obj.(*workItem)
+	if !ok {
+		m.logger.Fatalf("unexpected work item type: want: %s, got: %s\n", "*workItem", reflect.TypeOf(obj).Name())
+	}
+
+	ok, err := prober.Do(
+		item.context,
+		m.transportFactory(),
```

Comment:
with the newest updates if you use prober transport we can avoid passing transport factory around, but up to you.